test/jdk/tools/sincechecker/SinceChecker.java

```diff
@@ -0,0 +1,769 @@
+            } catch (IOException | NullPointerException e) {
+                wrongTagsList.add("module-info.java not found or couldn't be opened AND this module has no unqualified exports\n");
+            }
+        }
+        return version;
+    }
+
+    private String getPackageVersionFromFile(Path packagePath, ModuleElement.ExportsDirective ed) {
+        Path pkgInfo = packagePath.resolve(ed.getPackage()
+                        .getQualifiedName()
+                        .toString()
+                        .replaceAll("\\.", "/")
+                )
+                .resolve("package-info.java");
+
+        String packageTopVersion = null;
+        if (Files.exists(pkgInfo)) {
+            try {
+                byte[] packageAsBytes = Files.readAllBytes(pkgInfo);
+                String packageContent = new String(packageAsBytes, StandardCharsets.UTF_8);
+                packageTopVersion = extractSinceVersionFromText(packageContent).toString();
+            } catch (IOException | NullPointerException e) {
+                wrongTagsList.add("package-info.java not found or couldn't be opened\n");
+            }
+        }
+        return packageTopVersion;
+    }
+
+
+    private void analyzePackageCheck(PackageElement pe, JavacTask ct, EffectiveSourceSinceHelper javadocHelper) {
+        List<TypeElement> typeElements = ElementFilter.typesIn(pe.getEnclosedElements());
+        for (TypeElement te : typeElements) {
+            analyzeClassCheck(te, null, javadocHelper, ct.getTypes(), ct.getElements());
+        }
+    }
+
+    private void analyzeClassCheck(TypeElement te, String version, EffectiveSourceSinceHelper javadocHelper,
+                                   Types types, Elements elementUtils) {
+        String currentjdkVersion = String.valueOf(Runtime.version().feature());
+        Set<Modifier> classModifiers = te.getModifiers();
+        if (!(classModifiers.contains(Modifier.PUBLIC) || classModifiers.contains(Modifier.PROTECTED))) {
+            return;
+        }
+        checkElement(te.getEnclosingElement(), te, types, javadocHelper, version, elementUtils);
+        te.getEnclosedElements().stream().filter(element -> element.getModifiers().contains(Modifier.PUBLIC)
+                        || element.getModifiers().contains(Modifier.PROTECTED))
+                .filter(element -> element.getKind().isField()
+                        || element.getKind() == ElementKind.METHOD
+                        || element.getKind() == ElementKind.CONSTRUCTOR)
+                .forEach(element -> checkElement(te, element, types, javadocHelper, version, elementUtils));
+        te.getEnclosedElements().stream()
+                .filter(element -> element.getKind().isDeclaredType())
+                .map(TypeElement.class::cast)
+                .forEach(nestedClass -> analyzeClassCheck(nestedClass, currentjdkVersion, javadocHelper, types, elementUtils));
+    }
```

Comment:
more possible applications of those utility predicate methods I mentioned earlier.